ci-operator/step-registry/cucushift/upgrade/rhel-worker/cucushift-upgrade-rhel-worker-commands.sh

```diff
@@ -0,0 +1,248 @@
+            return 0
+        fi
+    done
+    if [[ ${wait_upgrade} -le 0 ]]; then
+        echo -e "Upgrade timeout on $(date "+%F %T"), exiting\n" && return 1
+    fi
+}
+
+# Check version, state in history
+function check_history() {
+    local version state
+    version=$(oc get clusterversion/version -o jsonpath='{.status.history[0].version}')
+    state=$(oc get clusterversion/version -o jsonpath='{.status.history[0].state}')
+    if [[ ${version} == "${TARGET_VERSION}" && ${state} == "Completed" ]]; then
+        echo "History check PASSED, cluster is now upgraded to ${TARGET_VERSION}" && return 0
+    else
+        echo >&2 "History check FAILED, cluster upgrade to ${TARGET_VERSION} failed, current version is ${version}, exiting" && return 1
+    fi
+}
+
+function check_mcp() {
+    local out updated updating degraded try=0 max_retries=30
+    while (( try < max_retries )); 
+    do
+        echo "Checking worker pool status #${try}..."
+        echo -e "oc get mcp\n$(oc get mcp)"
+        out="$(oc get mcp worker --no-headers)"
+        updated="$(echo "${out}" | awk '{print $3}')"
+        updating="$(echo "${out}" | awk '{print $4}')"
+        degraded="$(echo "${out}" | awk '{print $5}')"
+
+        if [[ ${updated} == "True" && ${updating} == "False" && ${degraded} == "False" ]]; then
+            echo "Worker pool status check passed" && return 0
+        fi
+        sleep 120
+        (( try += 1 ))
+    done
+    echo >&2 "Worker pool status check failed" && return 1
+}
+
+if [[ -f "${SHARED_DIR}/kubeconfig" ]] ; then
+    export KUBECONFIG=${SHARED_DIR}/kubeconfig
+fi
+
+# Setup proxy if it's present in the shared dir
+if [[ -f "${SHARED_DIR}/proxy-conf.sh" ]]; then
+    # shellcheck disable=SC1091
+    source "${SHARED_DIR}/proxy-conf.sh"
+fi
+
+# oc cli is injected from release:target
+run_command "which oc"
+run_command "oc version --client"
+run_command "oc get mcp"
+run_command "oc get machineconfig"
+
+export TARGET="${OPENSHIFT_UPGRADE_RELEASE_IMAGE_OVERRIDE}"
+TARGET_VERSION="$(env "NO_PROXY=*" "no_proxy=*" oc adm release info "${TARGET}" --output=json | jq -r '.metadata.version')"
+TARGET_MINOR_VERSION="$(echo "${TARGET_VERSION}" | cut -f2 -d.)"
+export TARGET_VERSION
+export TARGET_MINOR_VERSION
+echo -e "Target release version is: ${TARGET_VERSION}\nTarget minor version is: ${TARGET_MINOR_VERSION}"
+
+SOURCE_VERSION="$(oc get clusterversion --no-headers | awk '{print $2}')"
+SOURCE_MINOR_VERSION="$(echo "${SOURCE_VERSION}" | cut -f2 -d.)"
+export SOURCE_VERSION
+export SOURCE_MINOR_VERSION
+echo -e "Source release version is: ${SOURCE_VERSION}\nSource minor version is: ${SOURCE_MINOR_VERSION}"
+echo -e "The source release version is gotten from clusterversion resource, that can not stand for the current version of worker nodes!"
+
+if [[ $(oc get nodes -l node.openshift.io/os_id=rhel) != "" ]]; then
+    run_command "oc get node -owide"
+    rhel_repo
+    rhel_upgrade
+    check_mcp
+fi
+check_history
```

Comment:
should it move into if block?

Reply:
Yeah, it is okay, but leaving it here should be harmless too for safe guard.

Reply:
hmm, to those upgrade without rhel worker, this step `cucushift-upgrade-rhel-worker` should be skipped actually, right? but due to `cucushift-upgrade-rhel-worker` is added into upgrade chain, if leaving it there, then it sounds like the rhel step will be executed anyway due to `check_history`.

Reply:
Or we can add a check in the very beginning of this step if there's RHEL workers inside the cluster, if no, just exit with 0 to skip all the following commands.

Reply:
>  to those upgrade without rhel worker, this step cucushift-upgrade-rhel-worker should be skipped actually, right?  but due to cucushift-upgrade-rhel-worker is added into upgrade chain, if leaving it there, then it sounds like the rhel step will be executed anyway due to check_history.

Right, the step is always executed. For the cluster without rhel workers, `check_history` will be executed against cluster, just a few second, so I think it is harmless.

Reply:
If that confuses the reading of codes, I am okay to move it to if block. It is not a big deal.

Reply:
Done.